Configuration resolver

julienfalque opened the pull request:
Implements #2482 using `symfony/options-resolver` component.

Configurable fixers implement a new `ConfigurationDefinitionFixerInterface` that extends `ConfigurableFixerInterface` to add a `defineConfiguration()` method. This method must return an `OptionsResolver` instance that defines the configuration options. The interface is temporary (thus deprecated) and to be merged into `ConfigurableFixerInterface` in 3.0.

A few deprecations are introduced: some fixers can be configured by passing a list of strings, e.g. `['foo', 'bar']` but the `symfony/options-resolver` component is not designed for this. In such case, passing the list as the root configuration array is deprecated and a new dedicated option has been created to pass the list, e.g. `['option' => ['foo', 'bar']]`. This also allows to add new options to these fixers in the future. Fixers that handle such configuration will automatically convert the passed array and trigger a deprecation notice. Also, as `symfony/options-resolver` only deals with arrays, passing `NULL` to set the default configuration is deprecated, an empty array must be passed instead.

**Fix command help, before:**

![fix-help-old](https://cloud.githubusercontent.com/assets/1736542/22863087/a17da838-f13a-11e6-9991-68cc8d9562f2.png)

**Fix command help, after:**

![fix-help-new](https://cloud.githubusercontent.com/assets/1736542/22863088/a19e8134-f13a-11e6-80a2-209da5129814.png)

**Describe command, before:**

![describe-old](https://cloud.githubusercontent.com/assets/1736542/22863090/a6f5906e-f13a-11e6-8019-5b768cff43ec.png)

**Describe command, after:**

![describe-new](https://cloud.githubusercontent.com/assets/1736542/22863091/a9ebdc10-f13a-11e6-945a-9f84de1fdc4e.png)

See also the [new README file](https://github.com/julienfalque/PHP-CS-Fixer/blob/configuration-resolver/README.rst).

GrahamCampbell commented:
Should probably be targeted at master, not 2.0.

GrahamCampbell commented:
> With this PR, FixerInterface has 4 new methods

Adding methods to the interface is a major breaking change is it not?

julienfalque commented:
Woops, of course it is 😅

I reverted `ConfigurableFixerInterface` removal and added a new `ConfigurationDefinitionFixerInterface` with the new methods instead. Both interfaces are deprecated and to be removed and merged into `FixerInterface` in 3.0.

julienfalque commented:
By the way, does anyone know why AppVeyor queues all PHP-CS-Fixer's builds and never runs them?

julienfalque commented:
Rebased onto master as introducing deprecations requires a minor version incrementation.

keradus commented on README.rst:
```diff
@@ -182,480 +182,594 @@ automatically fix anything:
-   | PHP arrays should be declared using the configured syntax (requires PHP
-   | >= 5.4 for short syntax).
-   | *Rule is: configurable.*
+**array_syntax**
```
why dropping the bullet ?

julienfalque replied:
Because I was not sure that having the configuration options as a sublist would be nice, but I checked and it's ok, so I restored the bullets.

julienfalque commented:
I added configuration details in the `describe` command output. By the way, do you think I should add same colors for configuration details in the `fix` command help?

julienfalque commented:
I replaced `WhitespacesFixerConfig` by configuration options `indent` and `line_ending`. The `WhitespacesFixerConfig` class and all methods using it are now deprecated.

`FixerFactory::setWhitespacesConfig()` has no replacement yet. I'm thinking about something like a `setDefaultOptions()` method.

keradus commented:
> > I replaced WhitespacesFixerConfig by configuration options indent and line_ending. The WhitespacesFixerConfig class and all methods using it are now deprecated.

Please don't make such changes before alignment on it.
Big :-1: for that. One thing is that it is pointless to have this config specified for each fixer separately (repeating it for half of the fixers in docs is pointless), but bigger thing is or now one could configure different whitespaces for different fixers.
Please revert that change.

julienfalque commented:
> One thing is that it is pointless to have this config specified for each fixer separately (repeating it for half of the fixers in docs is pointless), but bigger thing is now one could configure different whitespaces for different fixers.

I didn't though about this and I completely agree. I removed the related commit.

The point was to have a single configuration entry point and I still think it would be great, but I'll try to figure out a better solution later, maybe in a separate PR.

keradus commented:
Thank you
Totally :+1: for not doing everything in one PR ;)

julienfalque commented:
Added description to all configuration options and tweaked the README.

Last thing I'd like to do is to deprecate `getConfigurationDescription()` and `getDefaultConfiguration()` methods from `FixerDefinitionInterface` since the details are also available through the new configuration definition. Do you agree?

keradus commented on src/AbstractFixer.php:
```diff
@@ -80,6 +86,63 @@ public function supports(\SplFileInfo $file)
+            @trigger_error(
+                'Passing NULL to set default configuration is deprecated and will not be supported in 3.0, use an empty array instead.',
+                E_USER_DEPRECATED
+            );
```
:-1: for this change. If we want to deprecate sth, I would rather split this method into 2 - one to set config, one to reset it

julienfalque replied:
Passing an empty array would still reset to default configuration anyway, and this is what this new method would do. I'm not sure adding it is relevant.

keradus replied:
If we want to have a deprecation let us make it the way it will follow separation of concerns. If not, then no need for introducing deprecation.

SpacePossum replied:
for completeness, as discussed @ https://github.com/julienfalque/PHP-CS-Fixer/pull/1#discussion_r102440937 this deprecation is indeed valid and should stay in this PR :+1:

keradus commented on src/Fixer/Alias/NoMixedEchoPrintFixer.php:
```diff
@@ -24,14 +22,9 @@
  * @author SpacePossum
  */
-final class NoMixedEchoPrintFixer extends AbstractFixer implements ConfigurableFixerInterface
+final class NoMixedEchoPrintFixer extends AbstractFixer
```
dropping interface is a BC breaker

julienfalque replied:
This is because the interface is implemented by `AbstractFixer` now, but I'm going to revert that change.

keradus commented on src/Fixer/ConfigurationDefinitionFixerInterface.php:
```diff
@@ -0,0 +1,42 @@
+     *
+     * @return bool
+     */
+    public function isConfigurable();
```
what for? if the fixer is not configurable, it must not follow this interface

julienfalque replied:
Because I wanted this interface to be merged in `FixerInterface` in 3.0 so that fixers use this method instead of an interface to tell whether they are configurable.

keradus replied:
Then the getter (`getConfigurationDefinition` or sth) could throw Exception (as it will do it even if `isConfigurable` will return `false`, as both methods are public). Is it not ?

julienfalque replied:
Yep, this method was useful only because the interface was implemented by all fixers via `AbstractFixer`. As I am reverting this, the method becomes useless and I'm going to remove it anyway.

julienfalque commented on src/AbstractFixer.php:
```diff
@@ -12,32 +12,38 @@
  * @internal
  */
-abstract class AbstractFixer implements FixerInterface, DefinedFixerInterface
+abstract class AbstractFixer implements FixerInterface, ConfigurationDefinitionFixerInterface, DefinedFixerInterface
```
Reverting.

julienfalque commented:
I reverted the implementation of `ConfigurationDefinitionFixer` in `AbstractFixer`. Now, only fixers that are actually configurable implement it so I removed the `isConfigurable()` method from it.

I also added tests on the internal `OptionsResolver` that extends Symfony's one.

GrahamCampbell commented:
Not sure how this simplifies things. There's twice as much code now...

GrahamCampbell commented:
That said, the new help output is very nice. :P

julienfalque commented:
@GrahamCampbell You're right, maybe some parts could be simplified (thinking about `OptionsResolver::setNormalizer()` usages for example). I'll have a look at it.

keradus commented:
side request: @juliendufresne , if possible, please avoid squashing.
it's hard to see the changes you made across the time if the commits get squashed and one need to read whole diff again.
thanks !

julienfalque commented:
@keradus I use `git commit --fixup` a lot, is it ok if I still push fixup commits so we can squash them before merging?

keradus commented:
sure, don't bother about squashing on your side. we do it automatically during merging anyway.
(the only exception is when there are many authors to not remove authorship)

julienfalque commented:
@keradus When I edit the original post of the PR, fabbot.io checks are triggered while the code didn't change, is it expected?

SpacePossum commented:
fabbot.io works in mysterious ways, it's not something to worry about ;)

keradus commented:
new github hook msg is send, so new processing is executed

keradus commented:
@SpacePossum , we have just release `v2@stable`, I don't want to have this a BC breaker for 3.0, it will make maintainable 2.x<=>3.x harder. I don't wanna have BC breakers for at least next few months and I believe this PR could be achieved without breaking BC.

keradus commented on src/FixerDefinition/FixerDefinition.php:
```diff
@@ -40,6 +40,15 @@ public function __construct(
+        } elseif (5 === func_num_args()) {
+            @trigger_error('Argument #5 of FixerDefinition::__construct() is deprecated and will be dropped in 3.0.', E_USER_DEPRECATED);
+        } else {
+            $riskyDescription = $configurationDescription;
```
Why `$riskyDescription = $configurationDescription` ?

keradus replied:
ah, as you try to remove not 5&6 arg, but 4&5

julienfalque commented:
> I believe this PR could be achieved without breaking BC

I can remove some of the deprecations introduced in this PR for the moment but I think they make sense. However, there is one that cannot be avoided because of `symfony/options-resolver` not handling some configuration cases as described in the OP.

I do understand the maintenance burden this PR may create so maybe it's wiser to wait before merging. But on the other hand, having full details of configuration options right in the README file seems valuable to me.

keradus commented:
@julienfalque, deprecations without breaking a BC are OK. Like triggering warnings or introducing new interfaces.
In my prev msg I refer to fact I remove the milestone, not to your code ;)

keradus commented:
@julienfalque, sorry for very small output from my side.

I think I do want to have here for sure is no BC breakers (deprecations and new interfaces are OK, as stated). For sure I would like to have more time and review whole PR once in the future, but for now it would be great if anyone from community would help reviewing this every-day bigger PR.

julienfalque commented:
@keradus Sure, no problem. I think I'm done here and I won't push more changes, unless I'm requested to. Anyway, thanks for the feedback you already gave :)

SpacePossum commented:
great @julienfalque :)
can you do the reviewers one more favor and rebase one more time, I hope to pick this one up shortly.

julienfalque commented:
Done.

SpacePossum commented on src/OptionsResolver.php:
```diff
@@ -0,0 +1,401 @@
+/**
+ * @internal
+ */
+class OptionsResolver extends BaseOptionsResolver
```
please make the class `final`

julienfalque commented:
Adressed most of the review comments. Note that with `symfony/phpunit-bridge` being used for tests, the tested code will **not** be allowed to throw deprecation notices that are not explicitly expected in the related tests anymore.

Last point: as @SpacePossum noticed, there is no upgrade path to avoid using code that is deprecated in this PR. While I agree it would be better to have one, I wonder if it's really worth it: the only deprecated code that we can't avoid using is `ConfigurationDefinitionFixerInterface`, which is an interface so using it does not actually trigger any deprecation notices. @SpacePossum suggested to remove the `@deprecated` tag for now and add it again just before releasing `3.0`, which we may forget to do. So, is it worth it?

keradus commented on src/Fixer/Alias/NoMixedEchoPrintFixer.php:
```diff
@@ -13,25 +13,20 @@
     /**
-     * @deprecated will not be publicly available on 3.0
-     */
-    public static $defaultConfig = array('use' => 'echo');
```
why this was removed ?

julienfalque replied:
I didn't see it was public, I'm gonna update the deprecation message instead.

keradus commented on src/Fixer/ConfigurationDefinitionFixerInterface.php:
```diff
@@ -0,0 +1,28 @@
+/**
+ * @deprecated Will be merged into ConfigurableFixerInterface in 3.0
+ */
+interface ConfigurationDefinitionFixerInterface extends ConfigurableFixerInterface
```
I'm not sure is it a good way of doing things...

We have:
- `ConfigurableFixerInterface`
- `ConfigurationDefinitionFixerInterface extends ConfigurableFixerInterface`

what we want is: later is newest form how we want to specif configuration, all should stop using original one
what we does: newly added interface is deprecated, so one should not use it. as one won't use it, he will still use original interface


We should not add interface, that is deprecated from the very beginning of its existence.
We could add new interface and deprecate existing one.

julienfalque replied:
I know, the point was to add a new method to `ConfigurableFixerInterface` but this is a BC break, hence the new interface deprecated because it will be merged into the current one in 3.0. I removed the `@deprecated` tag.

Current state is:
* `ConfigurableFixerInterface` not deprecated;
* `ConfigurationDefinitionFixerInterface` extends `ConfigurableFixerInterface` with a new method, not deprecated.

So now I can deprecate `ConfigurableFixerInterface`, making `ConfigurationDefinitionFixerInterface` the interface to use in 3.0. But I think `ConfigurableFixerInterface` is a better name, I would like to keep it in 3.0 and remove `ConfigurationDefinitionFixerInterface` instead. So what about having:

* `ConfigurableFixerInterface` (not ?) deprecated, will have a new method in 3.0;
* `ConfigurationDefinitionFixerInterface` extends `ConfigurableFixerInterface`, not deprecated, will be deprecated in 3.0 and removed in 4.0.

Is such process desirable? worth it? too complicated? Does it make sense to deprecate an interface while it is not to be removed later?

keradus replied:
@juliendufresne 

Please, don't go that far to think about v4, it won't be there for months, if not years.

current state:
We have 2 interfaces for same thing, user will have no idea which one he should use.
We should deprecate the one that we don't want to be used, here, it would be the original one.

Please:
- deprecate `ConfigurableFixerInterface`, as we don't want user to use it

What we will do in v4:
- drop deprecations, so we will remove `ConfigurableFixerInterface`
- we could make BC breakers, like renaming `ConfigurationDefinitionFixerInterface` (maybe into `ConfigurableFixerInterface`)

julienfalque replied:
Fair enough, I added the deprecation.

julienfalque commented:
I think I adressed all last reviews comment.

julienfalque commented on src/Fixer/Comment/HeaderCommentFixer.php:
```diff
@@ -13,63 +13,70 @@
-final class HeaderCommentFixer extends AbstractFixer implements ConfigurableFixerInterface, WhitespacesAwareFixerInterface
+final class HeaderCommentFixer extends AbstractFixer implements ConfigurationDefinitionFixerInterface, WhitespacesAwareFixerInterface
 {
-    const HEADER_PHPDOC = 'PHPDoc';
```
I will restore these constants. Removing them is not needed here and is a BC break anyway.

hkdobrev commented:
I'm not into the codebase of the PHP CS Fixer enough to help here, but the changes from this PR seems really good, so is there anything else needed here? I see AppVeyor is failing but seems like a temporary networking problem.

@julienfalque Thank you for all the work!

julienfalque commented:
Sadly AppVeyor isn't as reliable as Travis and is very slow. Wish we could run Travis on Windows and get rid of AppVeyor.

keradus commented on src/ConfigurationException/InvalidConfigurationException.php:
```diff
@@ -24,11 +24,16 @@
+     * @param \Exception|null $previous
      */
-    public function __construct($message, $code = null)
+    public function __construct($message, $code = null, \Exception $previous = null)
```
in `\Exception` definition, it is `\Throwable`, not `\Exception`

julienfalque replied:
Good catch, but won't it break on PHP 5.x? If so, this exception class is only supposed to wrap `OptionsResolver`'s exceptions, so it should not be a big issue.

keradus commented on src/Console/Command/ReadmeCommand.php:
```diff
@@ -216,9 +216,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
+        $help = preg_replace('#^ \* ``(.+)``(.*?\n)#m', "* **$1**$2\n", $help);
+        $help = preg_replace('#^   \\| #m', '  ', $help);
+        $help = preg_replace('#^   \\|#m', '', $help);
+        $help = preg_replace('#^(?=  \\*Risky rule.\\*$)#m', "\n", $help);
```
why such a change here and not in `getHelp` method ?

julienfalque replied:
Because I wanted to keep both `fix --help` output and README file nice. Which change would like to move exactly? Everything?

keradus replied:
(please freeze any update of readme file until further notice)

keradus replied:
readme is fine for me :+1:

keradus commented on src/OptionsResolver.php:
```diff
@@ -0,0 +1,401 @@
+/**
+ * @internal
+ */
+final class OptionsResolver extends BaseOptionsResolver
```
This class adds support for `rootConfigurationOption`.
Is it the only reasoning for this class ?

julienfalque replied:
No: the base `OptionsResolver` class does not provide all the getters I added. It also provides descriptions support.

keradus replied:
Sadly, I do not like the approach here. Don't get me wrong, I don't have better solution ( at least yet ), but copy-paste all private properties and duplicate the code to handle them is a pretty big boilerplate. Imagine that sb would fix a bug in original implementation - we would not be aware of this in our duplicated code.

`OptionsResolver` class is not final, but yet it uses private properties.
Could you propose to make the properties protected instead ?

keradus replied:
Another option would be not to inherit, but to compose.

Quick draft:

```php
final class FixerConfigurationResolver
{
    private $options; // FixerOptions[]
    private $optionsResolver; // Symfony class instance we use in following methods

    public function setRoot(string $name) { ... } // to be removed on 3.x line
    public function setOption(FixerOption $option) { ... }
    public function getOptions() :FixerOption[] { ... }
    public function resolve(array $opts) :array {...}
}

final class FixerOption // just a DTO
{
// required
    private $name;
    private $description;

// optional
    private $default;
    private $normalizer;
    private $validator;

    // getters, setters
}
```

that way we do not modify or rely on internal Symfony code.

julienfalque replied:
Yes, I really don't like this approach neither. Your idea seems way better, will do that.

keradus replied:
be aware that `FixerConfigurationResolver` will be part of public interface of this package, same for `FixerOption`.
Make them final with interfaces.

julienfalque replied:
Do we really need interfaces for that?

keradus replied:
If we have a interface for a fixer itself, then an interface for configuring it, and then we will have a non-interfaced final class it will break open principle here. Remember that everywhere here end-user may want to create his own Fixer (rule class). Then, he will use our DTO. Great, but he cannot modify the behaviour, if for some reason he would want to. Then, with interface, he could provide his own logic.

keradus replied:
This (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2500#discussion_r104912822) proposal would make a lot of code change in this PR (update of all configs definitions). Before doing that it would be great if one could double check the idea.

cc @SpacePossum , willing to take a look ?

keradus replied:
cc @SpacePossum , willing to take a look at my proposal (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2500#discussion_r104912822) ? (not necessarily whole PR)

SpacePossum replied:
sure thing, didn't find the time yet

julienfalque replied:
Implemented.

keradus commented on src/OptionsResolver.php:
```diff
@@ -0,0 +1,401 @@
+     *
+     * @return $this
+     */
+    public function addNormalizer($option, \Closure $normalizer)
```
this is used only once.
why can't we rely only on setNormalizer ?

julienfalque replied:
Don't remember. I'll have a look, maybe I can remove it.

keradus replied:
If the only purpose of this method is to use `unbind` helper then drop both of them.
Nice workaround, but we gonna drop 5.3 anyway

julienfalque replied:
I needed it to implement `setAllowedValueIsSubsetOf()`. I'll check if I can implement it differently.

keradus replied:
Could you elaborate? I don't see connection between those 2 functions:

```
λ grep -r addNormalizer src\Fixer  | wc -l
      1
λ grep -r setAllowedValueIsSubsetOf src\Fixer  | wc -l
      8
```

julienfalque replied:
`setAllowedValueIsSubsetOf()` defines a normalizer, so a subsequent call to `setNormalizer()` would override it, hence the `addNormalizer()` method. But I'll find a better way.

keradus replied:
wow, don't mix what is allowed with normalizer. it's different purpose.

replace `setAllowedValueIsSubsetOf` should be relaced with:
```php
->setAllowedTypes('opt', 'array')
->setAllowedValues('opt', function (...) {...})
```

julienfalque replied:
Fixed.

keradus commented on src/AbstractFixer.php:
```diff
@@ -26,6 +30,11 @@
 abstract class AbstractFixer implements FixerInterface, DefinedFixerInterface
 {
     /**
+     * @var array
```
actually, it's `Options`

keradus replied:
actually, it may be null

julienfalque replied:
It can be `null` but not `Options`, the result of `OptionsResolver::resolve()` is an array which can contain any kind of value. I can use `array<string, mixed>|null` here if you want.

julienfalque replied:
Done.

keradus commented on src/AbstractFixer.php:
```diff
@@ -34,7 +43,7 @@ public function __construct()
         if ($this instanceof ConfigurableFixerInterface) {
             try {
-                $this->configure(null);
+                $this->configure(array());
```
I pointed it here to show it.
`getConfigurationDefinition` must not create new instance all the time it is called. (Especially with new approach of creating bunch of DTO objects).

Currently, we generate new instances for one fixer when:
1. class is created, default configuration is applied
2. class is configured with user-provided config

No need in real tool run for doubling the object instantation.
Then, we also test it:
3. With every tests execution in `setUp`
4. With some tests execution when we provide custom config.

Also, applying the default config should be ultra fast. 

---------------

Proposal: lazy loading.
in whole src/fixers/*/ rename `getConfigurationDefinition` to `createConfigurationDefinition` and add new `getConfigurationDefinition` to this abstract class that will use `createConfigurationDefinition` if the instance was not yet created.

julienfalque replied:
I agree and tried to do that but I faced two issues.

First, it has unexpected results with configuration definitions that depend on some context of the fixer. It made `header_comment` configuration throw errors because it depends on the `WhitespacesFixerConfig` instance of the fixer, which can change between `configure()` calls. It can be fixed by forcing this fixer to always create a new configuration definition instance instead of reusing the first one.

But the second issue is an increasing of memory usage: I wasn't able to run the test suite without exceeding the memory limit on my setup. The limit is the default 128M so it's not high but the test suite usually requires ~40M. I'll have to investigate more to determine if this memory issue is a direct consequence of this change or something else that requires some optimization first. It looks like this only concerns the test suite though, as running the tool on the project didn't show a noticeable change in the consumed memory.

Anyway, applying this change didn't seem to have an impact on performance either, so I'm not sure it's worth it.

keradus replied:
memory issue could be figured out. let us focus on first one first.

> It made header_comment configuration throw errors because it depends on the WhitespacesFixerConfig instance of the fixer, which can change between configure() calls.

It should not be done that way.
It's possible to run both methods - `configure` and `setWhitespacesConfig`.
If one would call: `setWhitespacesConfig`, `configure` - it works.
If one would call: `setWhitespacesConfig`, `configure`, `setWhitespacesConfig`, `configure` - it works.
If one would call `configure`, `setWhitespacesConfig` - it doesn't work, because configuration is set to outdated, previous whitespaces - one need to call `configure` all after all times he would call `setWhitespacesConfig` for `header_comment` Fixer.
I hope I show you the shortcoming of current implementation.

Both methods must be independent. I think you must remove that normalizer and call it not during applying configuration , but during using the value from config

keradus replied:
issue with `header_comment` fixed.
could you investigate memory issue when remembering the definition ?

keradus replied:
2nd issue solved

keradus commented on src/Console/Command/CommandHelp.php:
```diff
@@ -269,6 +270,21 @@ public static function toString($value)
 
+    public static function getDisplayableAllowedValues(FixerOption $option)
+    {
+        if (null !== $allowed = $option->getAllowedValues()) {
```
please avoid side effects.
condition should not assign value

julienfalque replied:
Ah right, I forgot this one. I will assign the variable before the `if`. Just out of curiosity, why do you want to avoid such syntax? IMO it's more convenient and readable.

keradus replied:
good practice. too easy to write
`if `($foo = bar()) {}`
then you don't know did one wanted to just assign and check is it truly or compare function result to variable

julienfalque replied:
Yep, I only do this in case I have both assignment **and** comparison. In `if ('foo' === $bar = baz()) { }` there is no ambiguity.

mlocati replied:
I find far more readable the following code:
```php
$bar = baz();
if ('foo' === $bar) {
```

julienfalque replied:
Maybe we could write a fixer to force one way or the other 😋

julienfalque replied:
Assignment extracted.

keradus commented on src/Console/Command/CommandHelp.php:
```diff
@@ -269,6 +270,21 @@ public static function toString($value)
         return var_export($value, true);
     }
 
+    public static function getDisplayableAllowedValues(FixerOption $option)
```
what is the output ?

julienfalque replied:
PHPDoc added.

keradus commented on src/FixerConfiguration/FixerConfigurationResolver.php:
```diff
@@ -0,0 +1,116 @@
+     *
+     * @deprecated will be removed in 3.0
+     */
+    public function mapRootConfigurationTo($optionName)
```
this method shoud be extracted to separated class that will inherit from this class. that class should be not only deprecated but also internal. that way we wont produce new deprecated code already by dev users

julienfalque replied:
Actually this would require `FixerConfigurationResolver` to not be `final`, is that ok?

keradus replied:
no, let s think about it further

keradus commented on src/FixerConfiguration/FixerConfigurationResolver.php:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * @var array<string, FixerOptionInterface>
+     */
+    private $options = array();
```
we have an assoc array here. we don't need the keys that much. what about replacing it with `SplObjectStorage` (used as Set, not as Map) ?

julienfalque replied:
The keys are just used internally to easily know whether an option is defined using `array_key_exists()`. Not sure how using `SplObjectStorage` would be better, but why not?

keradus replied:
or, see next comment

keradus commented on src/FixerConfiguration/FixerConfigurationResolver.php:
```diff
@@ -0,0 +1,116 @@
+     *
+     * @return $this
+     */
+    public function addOption(FixerOptionInterface $option)
```
maybe make class immutable and pass options to constructor only ? wdyt?

julienfalque replied:
Having this method can be useful when a set of fixers share common options but some of them have specific ones too: the common options can be defined in an abstract parent class and the child can still add new options. I'd like to use that in #2220 :)

keradus replied:
btw: input could not be an array but any `iterable`

keradus commented on src/FixerConfiguration/FixerOption.php:
```diff
@@ -0,0 +1,225 @@
+     *
+     * @return \Closure
+     */
+    private function unbind(\Closure $closure)
```
let us drop this method. we dont care about 5.3

julienfalque replied:
Well, this method is actually useful on all PHP versions supported here... except PHP 5.3 which does not bind `$this` to `Closure`s by default 😁.

Maybe I can find another workaround though, or just get rid of the `setAllowedValueIsSubsetOf()` method but it's kind of convenient.

keradus replied:
ah, I misunderstand the code.
:+1: for unbinding

keradus commented on tests/ProjectCodeTest.php:
```diff
@@ -81,6 +83,15 @@ function (\ReflectionClass $interface) {
             'PhpCsFixer\Fixer\Import\OrderedImportsFixer' => array('sortingCallBack'),
+            'PhpCsFixer\FixerConfiguration\FixerConfigurationResolver' => array('addOption', 'mapRootConfigurationTo'),
+            'PhpCsFixer\FixerConfiguration\FixerOption' => array(
+                'removeDefault',
```
this one does not exist

julienfalque replied:
This was a method I wrote and then removed, but I forgot to remove it from here too.

keradus commented on tests/FixerConfiguration/FixerOptionTest.php:
```diff
@@ -0,0 +1,125 @@
+    public function testSetAllowedValueIsSubsetOf()
+    {
+        $option = new FixerOption('foo', 'Bar.');
+        $this->assertSame($option, $option->setAllowedValueIsSubsetOf(array('baz', 'qux')));
```
You check only fact that interface is fluent.
Not the logic of given method

keradus commented on src/Console/Command/CommandHelp.php:
```diff
@@ -270,9 +270,17 @@ public static function toString($value)
     {
-        if (null !== $allowed = $option->getAllowedValues()) {
+        $allowed = $option->getAllowedValues();
+        if (null !== $allowed) {
```
not only here, in other places as well

keradus commented on src/FixerConfiguration/FixerOption.php:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    /**
+     * @param string[]|string|null $allowedTypes
```
let us be strict here. consider this as PHP 7.1 code.
what type could you typehint in method prototype?
`?string $allowed` is cool, `?array $allowed` as well. but you cannot typehint `array|string|null $allowed`

julienfalque replied:
This is what Symfony OptionsResolver supports: `string|string[]`, thought we should just support the same. I added `null` as well but I'm not sure it's useful actually.

keradus replied:
no need for mimics symfony here. i prefer to be strict, then Zend releases PHP 7 which allows for that as well ;)
symfony code was created ages ago

julienfalque replied:
Ok, so just `string[]` (typehint `array`)?

keradus replied:
:+1:

keradus commented on src/FixerConfiguration/FixerOption.php:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    /**
+     * @param string[]|string|null $allowedValues
```
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2500/files#r105841300

keradus replied:
and then, you break this interface by passing function ;)

julienfalque replied:
I copy-pasted this from the previous one and forgot to update the PHPDoc. It should be `mixed` here.

keradus replied:
:-1: for mixed

julienfalque replied:
Why?

keradus replied:
`mixed` could be **anything**. eg `resource` (like img handler). Do you handle it here?

julienfalque replied:
Passing a `resource` does not make sense but it won't break validation, I'm fine with that. Do you want to be more strict anyway?

keradus replied:
strict is good
loosly is bad
preferably do what could be achieved in prototype in 7.1

keradus commented on tests/Fixer/LanguageConstruct/DeclareEqualNormalizeFixerTest.php:
```diff
@@ -72,13 +85,16 @@ public function provideCases()
 
     /**
+     * @param array $config
+     * @param mixed $expectedMessage
```
why mixed ?

keradus commented on src/FixerConfiguration/FixerOption.php:
```diff
@@ -0,0 +1,225 @@
+            ->setAllowedValues($this->unbind(function ($values) use ($option, $allowedArrayValues) {
+                foreach ($values as $value) {
+                    if (!in_array($value, $allowedArrayValues, true)) {
+                        throw new InvalidOptionsException(sprintf(
```
why not `return false` ?

julienfalque replied:
Would work too, but the error message would be something like _Option "foo" with value "array" is not valid_, which is less helpful that the error message here.

keradus replied:
Both msg provide same value

keradus commented on src/FixerConfiguration/FixerOption.php:
```diff
@@ -0,0 +1,225 @@
+     *
+     * @return $this
+     */
+    public function setAllowedValueIsSubsetOf(array $allowedArrayValues)
```
Sorry, I would totally drop it.
This method breaks so many things. It's a convenient helper, but it's nothing more. It doesn't provide any functionallity.
And it provides issues.

There are 2 concepts in `OptionsResolver` component:
- normalizers - which **modify** value (eg ensuring there is no trailing spaces in value)
- validators (`setAllowed...`) - which **check** value (eg if they are in set or passing function)

If I would like to set up few validators:
```php
$foo->setAllowedValues([
    [$array, $of, $allowed, $values],
    function ($value) { /* validates value */ },
]);
$foo->setAllowedValueIsSubsetOf([...]);
```

what is the effect? first two validators were lost with 3rd one.

One could try to collect all validators, but then those methods wouldn't be setters, they would be adders. I don't like that.
IMO, `setAllowedValueIsSubsetOf` should be removed.

keradus replied:
ref https://github.com/julienfalque/PHP-CS-Fixer/pull/2

keradus commented on src/Console/Command/CommandHelp.php:
```diff
@@ -279,16 +337,6 @@ function (FixerInterface $a, FixerInterface $b) {
-                $attributes[] = 'configurable';
-            }
-
             $description = wordwrap($description, 72, "\n   | ");
```
what is the difference between this function and newly introduced `wordwrap` method ?

julienfalque replied:
The added method does not include style tags when couting lines length.

keradus commented on src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php:
```diff
@@ -71,17 +55,27 @@ public function __construct()
+            ))
+        ;
+
+        return new FixerConfigurationResolverRootless('control_statements', array(
```
could be just `statements`

keradus commented on src/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixer.php:
```diff
@@ -29,28 +31,27 @@
+        'return',
+        'throw',
+        'use',
+        'useTrait',
```
is it possible to somehow make this deprecated in favor of `use_trait` ?

keradus commented on tests/FixerConfiguration/FixerConfigurationResolverTest.php:
```diff
@@ -0,0 +1,166 @@
+        } catch (InvalidOptionsException $catched) {
+        }
+
+        $this->assertSame($exception, $catched);
```
why not `expectException` instead of `$catched , try-catch and assertion` ?

julienfalque replied:
To check that the catched exception is exactly the thrown one, but I'm fine with `expectException()` if you want.

keradus replied:
Then I think the test is over engineered.  Aren't you testing OptionResolver of Symfony here ?

julienfalque replied:
No, this makes sure that `FixerConfigurationResolver::resolve()` does not replace/wrap the thrown exception. But yes, maybe not very important to test that, I'll use `setExpectedException()` instead.

keradus commented on README.rst:
```diff
@@ -197,482 +197,773 @@ automatically fix anything:
+
+  - ``align_double_arrow`` (``NULL``, ``false``, ``true``): whether to apply, remove or
+    ignore double arrows alignment; defaults to ``false``
+  - ``align_equals`` (``NULL``, ``false``, ``true``): whether to apply, remove or ignore
```
It should be `null, false, true`, not `NULL, false, true`.

PS yes, `var_export` is stupid

keradus commented on src/Console/Command/CommandHelp.php:
```diff
@@ -288,7 +288,15 @@ public static function getDisplayableAllowedValues(FixerOptionInterface $option)
+            usort(
+                $allowed,
+                function ($valueA, $valueB) use (&$toString) {
+                    return strcasecmp($toString($valueA), $toString($valueB));
```
why not just
```php
return `strcasecmp(CommandHelp::toString($valueA), CommandHelp::toString($valueB));`
```
?

julienfalque replied:
I guess I didn't drink enough coffee when writing this :p

keradus replied:
i just hate calling methods by string :D

julienfalque replied:
Agreed, this is a hacky workaround for PHP 5.3 that does not allow calling `self::foo()` in closures.

julienfalque replied:
Just didn't remember I could just replace `self` with the actual class name 😁.

keradus commented on src/AbstractFixer.php:
```diff
@@ -104,8 +110,12 @@ public function configure(array $configuration = null)
         }
 
+        if (null === $this->configurationDefinition) {
+            $this->configurationDefinition = $this->getConfigurationDefinition();
```
with this approach one will still call `getConfigurationDefitinion()` and create new object all the time.
it is getter that should provide cached object.

keradus commented on tests/FixerConfiguration/FixerOptionTest.php:
```diff
@@ -112,6 +116,10 @@ public function testGetNormalizers()
 
+        if (PHP_VERSION_ID >= 50400) {
+            $this->markTestIncomplete('Assertion below fails due to closure unbinding');
+        }
```
then, let us drop those assertions

keradus commented on src/FixerConfiguration/FixerOption.php:
```diff
@@ -50,13 +50,40 @@
         $this->name = $name;
         $this->description = $description;
+        $this->isRequired = $isRequired;
+        $this->default = $default;
```
```php
if ($isRequired && $default) {
    throw new...
}
```
?

keradus commented on src/FixerConfiguration/FixerOptionBuilder.php:
```diff
@@ -0,0 +1,126 @@
+
+namespace PhpCsFixer\FixerConfiguration;
+
+final class FixerOptionBuilder
```
I'm wondering... maybe internal ?

julienfalque replied:
Why? It could be useful to devs that write custom fixers.

keradus replied:
why they can't use `FixerOption` ?

julienfalque replied:
They can, just as we can too. But the builder is more convenient to avoid passing all the arguments to `FixerOption` constructor.

keradus commented on tests/FixerConfiguration/FixerOptionTest.php:
```diff
@@ -0,0 +1,100 @@
+        $this->assertSame(array('baz', 'qux'), $option->getAllowedValues());
+
+        if (PHP_VERSION_ID >= 50400) {
+            $this->markTestIncomplete('Assertion below fails due to closure unbinding');
```
then, let us drop those assertions

julienfalque replied:
Actually I rewrote similar assertions in another test to make them compatible, forgot to rewrite this one.

keradus commented on src/AbstractFixer.php:
```diff
@@ -110,7 +110,7 @@ public function configure(array $configuration = null)
         }
 
-        if (null === $this->configurationDefinition) {
+        if (null === $this->configurationDefinition || $this instanceof WhitespacesAwareFixerInterface) {
```
https://github.com/julienfalque/PHP-CS-Fixer/pull/7

keradus commented on src/AbstractFixer.php:
```diff
@@ -80,6 +95,44 @@ public function supports(\SplFileInfo $file)
+            $configuration = array();
+        }
+
+        if (null === $this->configurationDefinition || $this instanceof WhitespacesAwareFixerInterface) {
```
https://github.com/julienfalque/PHP-CS-Fixer/pull/7

keradus commented on src/AbstractFixer.php:
```diff
@@ -80,6 +95,44 @@ public function supports(\SplFileInfo $file)
+        }
+
+        if (null === $this->configurationDefinition || $this instanceof WhitespacesAwareFixerInterface) {
+            $this->configurationDefinition = $this->getConfigurationDefinition();
```
with this approach one will still call `getConfigurationDefitinion()` and create new object all the time.
it is getter that should provide cached object.

keradus commented on src/AbstractFixer.php:
```diff
@@ -80,6 +95,44 @@ public function supports(\SplFileInfo $file)
+
+        if (null === $this->configurationDefinition) {
+            $this->configurationDefinition = $this->getConfigurationDefinition();
+        }
```
with this approach one will still call `getConfigurationDefitinion()` and create new object all the time.
it is getter that should provide cached object.

keradus commented on src/FixerConfiguration/FixerOption.php:
```diff
@@ -0,0 +1,176 @@
+        $this->name = $name;
+        $this->description = $description;
+        $this->isRequired = $isRequired;
+        $this->default = $default;
```
for some reason GitHub shows it outdated, but it's not:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2500#discussion_r106368247

keradus replied:
resolved.

keradus commented on src/AbstractFixer.php:
```diff
@@ -41,9 +41,9 @@
+     * @var FixerConfigurationResolverInterface|null
      */
-    private $configurationDefinition;
+    protected $configurationDefinition;
```
Not sure if we want to split `AbstractFixer`. If extracting configurable part, one could also want to extract whitespaceConfigurable part. but then we would end up in 2 new abstract classes, and then `header_comment` want to inherit both of them...
Please rethink that, if we really want to move that part away, we should do it fully, like moving `configure` into `AbstractConfigurableFixer`, this property and handling in constructor

keradus commented on src/AbstractFixer.php:
```diff
@@ -26,15 +31,25 @@
+    /**
+     * @var FixerConfigurationResolverInterface|null
+     */
+    protected $configurationDefinition;
```
Not sure if we want to split `AbstractFixer`. If extracting configurable part, one could also want to extract whitespaceConfigurable part. but then we would end up in 2 new abstract classes, and then `header_comment` want to inherit both of them...
Please rethink that, if we really want to move that part away, we should do it fully, like moving `configure` into `AbstractConfigurableFixer`, this property and handling in constructor

keradus commented on src/Console/Command/CommandHelp.php:
```diff
@@ -298,9 +353,53 @@ function (FixerInterface $a, FixerInterface $b) {
-                sort($attributes);
-                $help .= sprintf("   | *Rule is: %s.*\n", implode(', ', $attributes));
+            if ($fixer->isRisky()) {
+                $help .= "   | *Risky rule.*\n";
```
maybe replace it with sth like `Rule may be risky: $descriptionFromDefinition` ?

keradus commented:
Thank you @julienfalque.

julienfalque commented:
Thanks everyone for helping me on this!

SpacePossum commented:
@julienfalque awesome work, great addition to the tool! :D  thanks so much!
